packages/next/build/webpack/plugins/middleware-source-maps-plugin.ts

```diff
@@ -0,0 +1,49 @@
+import { webpack } from 'next/dist/compiled/webpack/webpack'
+import type { webpack5 } from 'next/dist/compiled/webpack/webpack'
+
+/**
+ * Produce source maps for middlewares.
+ * Currently we use the same compiler for browser and middlewares,
+ * so we can avoid having the custom plugins if the browser source maps
+ * are emitted.
+ */
+export const getMiddlewareSourceMapPlugins = (params: {
+  isProductionBrowserSourceMapsOn: boolean
+}) => {
+  if (params.isProductionBrowserSourceMapsOn) {
+    return []
+  }
+
+  return [
+    new webpack.SourceMapDevToolPlugin({
+      filename: '[file].map',
+      include: [
+        // Middlewares are the only ones who have `server/pages/[name]` as their filename
+        /^server\/pages\//,
+        // All middleware chunks
+        /^server\/middleware-chunks\//,
+      ],
+    }),
+    new MiddlewareSourceMapsPlugin(),
+  ]
+}
+
+/**
+ * Produce source maps for middlewares.
+ * Currently we use the same compiler for browser and middlewares,
+ * so we can avoid having the custom plugins if the browser source maps
+ * are emitted.
+ */
+class MiddlewareSourceMapsPlugin {
+  apply(compiler: webpack5.Compiler): void {
+    const PLUGIN_NAME = 'NextJsMiddlewareSourceMapsPlugin'
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.buildModule.tap(PLUGIN_NAME, (module) => {
+        module.useSourceMap = module.layer === 'middleware'
+      })
+      compilation.hooks.runtimeModule.tap(PLUGIN_NAME, (module) => {
+        module.useSourceMap = module.layer === 'middleware'
+      })
```

Comment:
```suggestion
```

runtimeModules are never in `middleware` layer, but maybe in middleware chunks. I would juest omit that and leave source maps on for them.